Cython/Compiler/ExprNodes.py

```diff
@@ -3246,7 +3287,34 @@ def infer_type(self, env, iterator_type=None):
             return fake_index_node.infer_type(env)
 
     def analyse_types(self, env):
+        def remove_const(item_type):
+            """
+            Removes the constness of a given type and its underlying templates
+            if any.
+
+            This is to solve the compilation error when the temporary variable used to
+            store the result of an iterator cannot be changed due to its constness.
+            For example, the value_type of std::map, which will also be the type of
+            the temporarry variable, is std::pair<const Key, T>. This means the first
+            component of the variable cannot be reused to store the result of each
+            iteration, which leads to a compilation error.
+            """
+            if item_type.is_const:
+                item_type = item_type.cv_base_type
+            if isinstance(item_type, CTypedefType):
+                item_type = item_type.typedef_base_type
```

Comment:
I don't know if this should be in a `while` loop? I think that `item_type` out of a typedef could probably be const.

Reply:
I'm not sure if item_type refers to the type of temporary variable that is reused inside a loop to store the result of an iterator for each iteration. When I tried to iterate std::map it caused a compilation error due to item_type being std::map<K, V>::value_type, which deduces to std::pair<const K, V>. Not sure if there is a better way to deal with this.

Reply:
Ah I don't think I explained very well. I was proposing something like

```
while True:
    item_type_in = item_type
    if item_type.is_const:
        item_type = item_type.cv_base_type
    if isinstance(item_type, CTypedefType):
        item_type = item_type.typedef_base_type
    if item_type == item_type_in: break
```

I just thought that you weren't guaranteed to remove all the consts in one go